rust/agama-dbus-server/src/questions/answers.rs

```diff
@@ -0,0 +1,291 @@
+                    )])),
+                    text: None,
+                    answer: "Ok2".to_string(),
+                    password: None,
+                },
+            ],
+        }
+    }
+
+    #[test]
+    fn test_class_match() {
+        let answers = get_answers();
+        let question = GenericQuestion {
+            id: 1,
+            class: "without_data".to_string(),
+            text: "JFYI we will kill all bugs during installation.".to_string(),
+            options: vec!["Ok".to_string(), "Cancel".to_string()],
+            default_option: "Cancel".to_string(),
+            data: HashMap::new(),
+            answer: "".to_string(),
+        };
+        assert_eq!(Some("Ok".to_string()), answers.answer(&question));
+    }
+
+    #[test]
+    fn test_no_match() {
+        let answers = get_answers();
+        let question = GenericQuestion {
+            id: 1,
+            class: "non-existing".to_string(),
+            text: "Hard question?".to_string(),
+            options: vec!["Ok".to_string(), "Cancel".to_string()],
+            default_option: "Cancel".to_string(),
+            data: HashMap::new(),
+            answer: "".to_string(),
+        };
+        assert_eq!(None, answers.answer(&question));
+    }
+
+    #[test]
+    fn test_with_password() {
+        let answers = get_answers();
+        let question = GenericQuestion {
+            id: 1,
+            class: "without_data".to_string(),
+            text: "Please provide password for dooms day.".to_string(),
+            options: vec!["Ok".to_string(), "Cancel".to_string()],
+            default_option: "Cancel".to_string(),
+            data: HashMap::new(),
+            answer: "".to_string(),
+        };
+        let with_password = WithPassword {
+            password: "".to_string(),
+            base: question,
+        };
+        let expected = (Some("Ok".to_string()), Some("testing pwd".to_string()));
+        assert_eq!(expected, answers.answer_with_password(&with_password));
+    }
+
+    /// An Answer matches on *data* if all its keys and values are in the GenericQuestion *data*.
+    /// The GenericQuestion can have other *data* keys.
+    #[test]
+    fn test_partial_data_match() {
+        let answers = get_answers();
+        let question = GenericQuestion {
+            id: 1,
+            class: "with_data".to_string(),
+            text: "Hard question?".to_string(),
+            options: vec!["Ok2".to_string(), "Maybe".to_string(), "Cancel".to_string()],
+            default_option: "Cancel".to_string(),
+            data: HashMap::from([
+                ("data1".to_string(), "value1".to_string()),
+                ("data2".to_string(), "value2".to_string()),
+                ("data3".to_string(), "value3".to_string()),
+            ]),
+            answer: "".to_string(),
+        };
+        assert_eq!(Some("Maybe".to_string()), answers.answer(&question));
+    }
+
+    #[test]
+    fn test_full_data_match() {
```

Comment:
I think this test does the same thing as `partial_data_match`. What is the point?

Reply:
well, difference here is that first test match only one answer. But here data2 match one answer and data1 another. So it test if wins correct answer.

Reply:
I see, maybe I can express it slightly better in code or comments, will try